src/platform/plugins/shared/discover/public/application/main/state_management/redux/internal_state.ts

```diff
@@ -139,8 +114,19 @@ export const internalStateSlice = createSlice({
         {}
       );
       state.tabs.allIds = action.payload.allTabs.map((tab) => tab.id);
+      state.tabs.recentlyClosedTabsById = action.payload.recentlyClosedTabs.reduce<
+        Record<string, RecentlyClosedTabState>
+      >(
+        (acc, tab) => ({
+          ...acc,
+          [tab.id]: tab,
+        }),
+        {}
+      );
+      state.tabs.recentlyClosedTabIds = action.payload.recentlyClosedTabs.map((tab) => tab.id);
       state.tabs.unsafeCurrentId = action.payload.selectedTabId;
-      state.tabs.recentlyClosedTabs = action.payload.recentlyClosedTabs;
+      state.persistedDiscoverSession =
+        action.payload.updatedDiscoverSession ?? state.persistedDiscoverSession;
```

Comment:
I updated the `setTabs` reducer to also accept an `updatedDiscoverSession`. This allows us to set the updated tabs state and persisted session during the same state update, and in the same tick that the runtime state is updated in the corresponding `setTabs` thunk. This eliminates any timing issues and ensures runtime states stay in sync, as well as giving us access to the updated session in the local persistence middleware.